src/main/java/org/folio/rest/impl/TenantPermsAPI.java

```diff
@@ -62,19 +77,396 @@ public void postTenantpermissions(OkapiPermissionSet entity, Map<String, String>
+  }
+
+  private List<OkapiPermission> getNewPerms(Map<String, Permission> dbPerms,
+      List<OkapiPermission> perms) {
+    List<OkapiPermission> newPerms = new ArrayList<>();
+    
+    if(perms != null) {
+      perms.stream()
+        .filter(perm -> perm.getPermissionName() != null)
+        .filter(perm -> !dbPerms.containsKey(perm.getPermissionName()))
+        .filter(perm -> {
+          // filter out renamed perms, we'll deal with them separately
+          for (String oldName : perm.getRenamedFrom()) {
+            if (dbPerms.containsKey(oldName)) {
+              return false;
+            }
+          }
+          return true;
+        })
+        .forEach(newPerms::add);
+    }
+
+    return newPerms;
+  }
+
+  private Map<OkapiPermission, Permission> getRenamedPerms(Map<String, Permission> dbPerms,
+      List<OkapiPermission> perms) {
+    Map<OkapiPermission, Permission> renamedPerms = new HashMap<>();
+    
+    if(perms != null) {
+      perms.stream()
+        .filter(perm -> perm.getPermissionName() != null)
+        .filter(perm -> !dbPerms.containsKey(perm.getPermissionName()))
+        .forEach(perm -> {
+          for (String oldName : perm.getRenamedFrom()) {
+            Permission dbPerm = dbPerms.get(oldName);
+            if (dbPerm != null) {
+              renamedPerms.put(perm, dbPerm);
+              break;
+            }
+          }
+        });
+    }
+    return renamedPerms;
+  }
+
+  private List<OkapiPermission> getNewAndModifiedPerms(ModuleId moduleId, Map<String, Permission> dbPerms,
+      List<OkapiPermission> okapiPerms) {
+    List<OkapiPermission> newPerms = getNewPerms(dbPerms, okapiPerms);
+    List<OkapiPermission> modifiedPerms = getModifiedPerms(moduleId, dbPerms, okapiPerms);
+    newPerms.addAll(modifiedPerms);
+    return newPerms;
+  }
+
+  private List<OkapiPermission> getModifiedPerms(ModuleId moduleId, Map<String, Permission> dbPerms,
+      List<OkapiPermission> okapiPerms) {
+    List<OkapiPermission> modifiedPerms = new ArrayList<>();
+
+    if(okapiPerms != null) {
+      okapiPerms.stream()
+        .filter(okapiPerm -> okapiPerm.getPermissionName() != null)
+        .filter(okapiPerm -> {
+            String name = okapiPerm.getPermissionName();
+            return dbPerms.containsKey(name)
+                && !comparePerms(okapiPerm, moduleId.getProduct(), dbPerms.get(name));
+          })
+        .forEach(modifiedPerms::add);
+    }
+    return modifiedPerms;
+  }
+
+  private List<Permission> getRemovedPerms(Map<String, Permission> dbPerms,
+      List<OkapiPermission> perms) {
+    List<Permission> removedPerms = new ArrayList<>();
+    List<String> permNames = new ArrayList<>();
+
+    if (perms != null) {
+      perms.stream()
+        .filter(perm -> perm.getPermissionName() != null)
+        .forEach(perm -> permNames.add(perm.getPermissionName()));
+    }
+
+    dbPerms.values().stream()
+      .filter(dbPerm -> !permNames.contains(dbPerm.getPermissionName()))
```

Comment:
This will include those renamed perms again. No harm but not efficient. Just a FYI